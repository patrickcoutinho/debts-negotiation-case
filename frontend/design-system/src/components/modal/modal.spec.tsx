import React from 'react';
import Modal from './modal';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';
import { faker } from '@faker-js/faker';
import { render, RenderResult, screen } from '@testing-library/react';

type SubjectProps = {
  openButton?: string;
  modalTitle?: string;
};

type SubjectTypes = {
  subject: RenderResult;
};

const makeSubject = ({
  openButton = faker.word.verb(),
  modalTitle = faker.word.verb(),
}: SubjectProps): SubjectTypes => {
  const subject = render(
    <Modal
      title={modalTitle}
      text="modal text"
      openButton={openButton}
      button="open"
      buttonAction={jest.fn()}
    />
  );

  return { subject };
};

describe('Modal', () => {
  test('should render Modal correctly', async () => {
    const {
      subject: { container },
    } = makeSubject({ openButton: 'Open' });

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  });

  test('should render Modal open button', async () => {
    const openButton = faker.word.verb();

    makeSubject({ openButton });

    expect(await screen.findByText(openButton)).toBeInTheDocument();
  });

  test('should open Modal on click button', async () => {
    const openButton = faker.word.verb();
    const modalTitle = faker.lorem.sentence();

    makeSubject({ openButton, modalTitle });

    const button = await screen.findByText(openButton);

    expect(
      screen.queryByText(new RegExp(modalTitle, 'i'))
    ).not.toBeInTheDocument();

    userEvent.click(button);

    expect(
      await screen.findByText(new RegExp(modalTitle, 'i'))
    ).toBeInTheDocument();
  });
});
