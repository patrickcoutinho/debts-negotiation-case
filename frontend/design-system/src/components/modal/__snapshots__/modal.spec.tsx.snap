// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render Modal correctly 1`] = `
<div>
  <button
    class="chakra-button css-63eb9s"
    type="button"
  >
    Open
  </button>
</div>
`;
